### PR TITLE
[REVIEW] Handle offsets for cython Column class

### PR DESCRIPTION
**Pull request by kkraus14**
This PR ballooned up a bit, apologies

Few things to-do still:

- [x] For columns with children (currently only strings and possibly categorical), is the offset only stored at the top level column? If so we need to build children accessors that are offset aware instead of just directly accessing the children objects.
- [x] Change `column_view_from_column` to return buffers in addition to the `gdf_column*` to prevent possible dangling pointers.
- [x] Throw on the data and mask Buffer accessors in there's a non-zero offset?
- [x] Create function to create a cython `Column` from a `unique_ptr<cudf::column_view>`

### CODE REVIEW

**Comment by shwina** [comment]
:fire:

Here's what I had for `from_column_view` in case it's useful. It constructs a `Column` from a `column_view`, additionally accepting an arbitrary Python which becomes the "owner" for the Buffers of the new Column. It's important that the `owner` be an arbitrary Python object (or at least, not limited to just Column), to support APIs like `contiguous_split`, where the owner will be a `DeviceBuffer`.

https://github.com/shwina/cudf/blob/6ae98111d305b6f98178bcf4bad53dc3e9a902b8/python/cudf/cudf/_libxx/column.pyx#L280-L317

File: python/cudf/cudf/_libxx/column.pyx [hunk 1]
```diff
@@ -309,38 +309,49 @@ cdef class Column:
         return build_column(data, dtype=dtype, mask=mask, children=children)
 
     @staticmethod
-    cdef Column from_column_view(column_view cv, Column owner):
+    cdef Column from_column_view(column_view cv, object owner):
```
Comment (shwina) [inline python/cudf/cudf/_libxx/column.pyx]: LGTM. Some documentation here on the behaviour would be useful down the road

File: python/cudf/cudf/core/column/datetime.py [hunk 2]
```diff
@@ -222,16 +225,16 @@ def default_na_value(self):
                 "datetime column of {} has no NaN value".format(self.dtype)
             )
 
-    def fillna(self, fill_value, inplace=False):
+    def fillna(self, fill_value):
```
Comment (kkraus14) [inline python/cudf/cudf/core/column/datetime.py]: Add back inplace parameter?

File: python/cudf/cudf/core/column/numerical.py [hunk 3]
```diff
@@ -150,10 +151,12 @@ def as_datetime_column(self, dtype, **kwargs):
...
 
     def as_numerical_column(self, dtype, **kwargs):
+        casted = libcudf.typecast.cast(self, dtype)
         return column.build_column(
-            data=libcudf.typecast.cast(self, dtype).data,
+            data=casted.data,
             dtype=np.dtype(dtype),
-            mask=self.mask,
+            mask=casted.mask,
+            size=casted.size,
```
Comment (kkraus14) [inline python/cudf/cudf/core/column/numerical.py]: add `offset=casted.offset`

File: python/cudf/cudf/core/column/numerical.py [hunk 4]
```diff
@@ -312,7 +315,7 @@ def find_and_replace(self, to_replace, replacement, all_nan):
         )
         return output
 
-    def fillna(self, fill_value, inplace=False):
+    def fillna(self, fill_value):
```
Comment (kkraus14) [inline python/cudf/cudf/core/column/numerical.py]: add inplace parameter back here?

File: python/cudf/cudf/core/column/string.py [hunk 5]
```diff
@@ -455,21 +455,54 @@ def __init__(self, mask=None, offset=0, children=()):
...
+                offsets_column = self.base_children[0].copy(deep=False)
+
+                # Shift offsets column by the parent offset.
+                offsets_column.offset = self.offset
+                offsets_column.size = self.size + 1
+
+                # Shift the chars offset by the new first element of the
+                # offsets column
+                chars_offset = offsets_column[0]
+                chars_size = offsets_column[self.size]
```
Comment (kkraus14) [inline python/cudf/cudf/core/column/string.py]: Need to do this after `offsets_column` is recalculated

File: python/cudf/cudf/core/column/string.py [hunk 6]
```diff
@@ -745,7 +780,7 @@ def find_and_replace(self, to_replace, replacement, all_nan):
                 " single values"
             )
 
-    def fillna(self, fill_value, inplace=False):
+    def fillna(self, fill_value):
```
Comment (kkraus14) [inline python/cudf/cudf/core/column/string.py]: add inplace parameter back?

File: python/cudf/cudf/core/series.py [hunk 7]
```diff
@@ -1145,7 +1149,7 @@ def __neg__(self):
 
     @property
     def cat(self):
-        return self._column.cat()
+        return self._column.cat(column_owner=self)
```
Comment (kkraus14) [inline python/cudf/cudf/core/series.py]: `column_owner` --> `parent`

File: python/cudf/cudf/_libxx/column.pyx [hunk 8]
```diff
@@ -59,29 +66,17 @@ cdef class Column:
...
-            raise TypeError("Expected a Buffer for data, got " +
-                            type(data).__name__)
-        if mask is not None and not isinstance(mask, Buffer):
-            raise TypeError("Expected a Buffer for mask, got " +
-                            type(mask).__name__)
-        if not pd.api.types.is_integer(size):
-            raise TypeError("Expected an integer for size, got " +
-                            type(size).__name__)
-        if not pd.api.types.is_integer(offset):
+    def __init__(self, data, size, dtype, mask=None, offset=None, children=()):
```
Comment (shwina) [inline python/cudf/cudf/_libxx/column.pyx]: Can't we just default `offset=0`? We set it to `0` below anyway.

**Comment by shwina** [comment]
Just need to default `offset=0` instead of `None` and I think we're good

File: python/cudf/cudf/core/series.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (kkraus14) [reply python/cudf/cudf/core/series.py]: I think this was updated already and you got some github cache. This reads as:
```
    @property
    def cat(self):
        return self._column.cat(parent=self)
```

File: python/cudf/cudf/_libxx/column.pyx [hunk 9]
```diff
@@ -162,7 +162,13 @@ cdef class Column:
...
     def mask(self):
         if self._mask is None:
-            if self.offset == 0 or self.base_mask is None:
+            if self.base_mask is None or (
+                self.offset == 0
+                and self.base_mask.size == calc_chunk_size(
+                    self.size,
+                    mask_bitsize
+                )
+            ):
```
Comment (shwina) [inline python/cudf/cudf/_libxx/column.pyx]: When does this come up? Maybe I'me missing something obvious..

Comment (kkraus14) [reply python/cudf/cudf/_libxx/column.pyx]: If the offset is 0 and the mask is already sized properly according to the column size then we don't want to go and create a newly allocated mask.

This is to guard against something like having a column of size 100 with 0 offset and slicing it down to the first 5 elements, where the new offset is 0 and the new size is 5. The mask should now be smaller than previously so we need to potentially calculate a new mask from the base_mask. But say you slice it from size 100 to size 99, the mask size stays the same so we don't need to calculate a new one and can just return a reference to the base_mask.

There's definitely some future optimizations possible here along the lines of:
- Just resizing the mask when possible (size is reduced but offset stayed 0)
- Using pointer arithmetic if offset is a multiple of 8 and the mask is large enough to accommodate the size properly.

**Comment by kkraus14** [comment]
Looking into the failures...

File: python/cudf/cudf/_libxx/column.pyx (hunk 9 above)
Comment (kkraus14) [reply python/cudf/cudf/_libxx/column.pyx]: Connected offline and there's a known issue here that we'll address in a follow up PR: cuDF Python doesn't currently guarantee the mask is byte padded to 64 bytes (or even 4 bytes) and this is known to break the `copy_bitmask` function.

**Pull request merged**